### PR TITLE
tests: do not check for invalid k/m combinations

### PR DESCRIPTION
**Pull request by tchaikov**
- [x] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

Fixes: http://tracker.ceph.com/issues/16500

### CODE REVIEW

**Comment by tchaikov** [comment]
tested using:

```
$ bin/ceph_erasure_code_benchmark --plugin lrc --workload encode --iterations 1 --size 4096 --parameter k=4 --parameter m=2 --parameter l=3 --erasure-code-dir lib
0.000069        4

$ bin/ceph_erasure_code_benchmark --plugin jerasure --workload encode --iterations 1 --size 4096 --parameter technique=liber8tion --parameter k=2 --parameter m=2 --parameter l=3 --erasure-code-dir lib/
0.000160        4

$ bin/ceph_erasure_code_benchmark --plugin jerasure --workload encode --iterations 1 --size 4096 --parameter technique=liber8tion --parameter k=1 --parameter m=2 --erasure-code-dir lib/
load: jerasure k=1 must be >= 2
```


before 1446972
```
$ bin/ceph_erasure_code_benchmark --plugin jerasure --workload encode --iterations 1 --size 4096 --parameter technique=liber8tion1 --parameter k=2 --parameter m=2 --erasure-code-dir lib/
load: jerasure
2018-11-12 10:59:39.049 7f2f689ac640 -1 ErasureCodePluginJerasure: technique=liber8tion1 is not a valid coding technique.  Choose one of the following: reed_sol_van, reed_sol_r6_op, cauchy_orig, cauchy_good, liberation, blaum_roth, liber8tion
```
after  1446972
```
$ bin/ceph_erasure_code_benchmark --plugin jerasure --workload encode --iterations 1 --size 4096 --parameter technique=liber8tion1 --parameter k=2 --parameter m=2 --erasure-code-dir lib/
load: jerasure technique=liber8tion1 is not a valid coding technique.  Choose one of the following: reed_sol_van, reed_sol_r6_op, cauchy_orig, cauchy_good, liberation, blaum_roth, liber8tion
```

File: src/erasure-code/jerasure/ErasureCodePluginJerasure.cc [hunk 1]
```diff
@@ -54,11 +54,10 @@ int ErasureCodePluginJerasure::factory(const std::string& directory,
     } else if (t == "liber8tion") {
       interface = new ErasureCodeJerasureLiber8tion();
     } else {
-      derr << "technique=" << t << " is not a valid coding technique. "
+      *ss << "technique=" << t << " is not a valid coding technique. "
```
Comment (majianpeng) [inline src/erasure-code/jerasure/ErasureCodePluginJerasure.cc]: @tchaikov . Should we check ss != nullptr?

Comment (tchaikov) [reply src/erasure-code/jerasure/ErasureCodePluginJerasure.cc]: no, i don't think so. please see `ErasureCode::sanity_check_k()`. FWIW, i think the reason why we have `ss` as a pointer here, is merely that `ss` is not a `const` here, and we have a tradition to pass non-const parameter as pointer.

**Comment by tchaikov** [comment]
retest this please.

**Comment by shylesh** [comment]
@tchaikov,

I did some basic testing on this PR and looks like it works as per your explanation of minimal lrc and other legitimate cases.

However I found some border case issues where few boundary conditions were not checked though its trivial.


./bin/ceph_erasure_code_benchmark  --plugin lrc --workload encode --iterations 400 --size 4096 --parameter k=2 --parameter m=0 --parameter l=1 --erasure-code-dir lib/
load: lrc load: jerasure k=1 must be >= 2


./bin/ceph_erasure_code_benchmark  --plugin lrc --workload encode --iterations 400 --size 4096 --parameter k=4 --parameter m=2 --parameter l=0 --erasure-code-dir lib/
*** Caught signal (Floating point exception) **
 in thread 7f0dedf8c640 thread_name:ceph_erasure_co
 ceph version 14.0.1-741-g1f44680034 (1f446800345d451a851d48d6c389651b78cc8085) nautilus (dev)
 1: (()+0x101634) [0x56057e2b0634]
 2: (()+0x12890) [0x7f0de38d4890]
 3: (ErasureCodeLrc::parse_kml(std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > >&, std::ostream*)+0x3d5) [0x7f0ddd705599]
 4: (ErasureCodeLrc::init(std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > >&, std::ostream*)+0x58) [0x7f0ddd707478]
 5: (ErasureCodePluginLrc::factory(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > >&, std::shared_ptr<ceph::ErasureCodeInterface>*, std::ostream*)+0x65) [0x7f0ddd6febeb]
 6: (ceph::ErasureCodePluginRegistry::factory(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > >&, std::shared_ptr<ceph::ErasureCodeInterface>*, std::ostream*)+0xe9) [0x56057e2b4c51]
 7: (ErasureCodeBench::encode()+0xfe) [0x56057e25d0fe]
 8: (ErasureCodeBench::run()+0x44) [0x56057e25cff0]
 9: (main()+0x7b) [0x56057e25e678]
 10: (__libc_start_main()+0xe7) [0x7f0de2b45b97]
 11: (_start()+0x2a) [0x56057e24ac1a]
2018-11-13 13:55:21.289 7f0dedf8c640 -1 *** Caught signal (Floating point exception) **


Thanks,
Shylesh

**Comment by tchaikov** [comment]
@shylesh fixed and repushed.

File: src/erasure-code/lrc/ErasureCodeLrc.cc [hunk 2]
```diff
@@ -324,7 +324,7 @@ int ErasureCodeLrc::parse_kml(ErasureCodeProfile &profile,
     }
   }
 
-  if ((k + m) % l) {
+  if (l == 0 || (k + m) % l) {
```
Comment (tchaikov) [inline src/erasure-code/lrc/ErasureCodeLrc.cc]: i could define a new error code for it, but the error message also applies to this case, i guess.

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2018-11-14_07:46:43-rados-wip-kefu-testing-2018-11-14-1049-distro-basic-smithi/
- http://pulpito.ceph.com/kchai-2018-11-14_23:26:43-rados-wip-kefu-testing-2018-11-15-0204-distro-basic-smithi/
- http://pulpito.ceph.com/kchai-2018-11-15_04:23:30-rados-wip-kefu-testing-2018-11-15-0204-distro-basic-smithi/
- http://pulpito.ceph.com/kchai-2018-11-15_05:33:10-rados-wip-kefu-testing-2018-11-15-0204-distro-basic-mira/

the test failures are fixed by
- #25106
- #25108

**Comment by tchaikov** [comment]
changelog

- rebased onto master.

**Comment by tchaikov** [comment]
http://pulpito.ceph.com/kchai-2018-11-15_09:38:15-rados-wip-kefu-testing-2018-11-15-1539-distro-basic-smithi/

**Comment by tchaikov** [comment]
retest this please.

**Pull request merged**